core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala

```diff
@@ -266,11 +273,14 @@ object ContainerPool {
    * since this would be picked up earlier in the scheduler and the container reused.
    *
    * @param pool a map of all free containers in the pool
+   * @param checkConcurrent if true only consider containers not processing other concurrent activations (otherwise disregard concurrent activation counts)
    * @return a container to be removed iff found
    */
-  protected[containerpool] def remove[A](pool: Map[A, ContainerData]): Option[A] = {
+  protected[containerpool] def remove[A](pool: Map[A, ContainerData], checkConcurrent: Boolean): Option[A] = {
+    // Try to find a Free container that does NOT have any active activations AND is initialized with any OTHER action
     val freeContainers = pool.collect {
-      case (ref, w: WarmedData) => ref -> w
+      case (ref, w: WarmedData) if !checkConcurrent || ActivationCounter.getOrElse(ref, 0) == 0 =>
```

Comment:
This becomes

```scala
case (ref, w: WarmedData) if w.activationCount == 0 =>
```

And you can remove the boolean flag if we can add the counter to `WarmedData`.